packages/cli/src/commands/devHandler.js

```diff
@@ -23,6 +24,71 @@ export const handler = async ({
 }) => {
   const rwjsPaths = getPaths()
 
+  // Starting values of ports from config (redwood.toml)
+  let apiPreferredPort = parseInt(getConfig().api.port)
+  let webPreferredPort = parseInt(getConfig().web.port)
+
+  // Assume we can have the ports we want
+  let apiAvailablePort = apiPreferredPort
+  let apiPortChangeNeeded = false
+  let webAvailablePort = webPreferredPort
+  let webPortChangeNeeded = false
+
+  // Check api port
+  if (side.includes('api')) {
+    apiAvailablePort = await getFreePort(apiPreferredPort)
+    if (apiAvailablePort === -1) {
+      errorTelemetry(
+        process.argv,
+        `Error could not determine a free port for the api server`
+      )
+      console.error(`Error could not determine a free port for the api server`)
+      process.exit(-1)
+    }
+    apiPortChangeNeeded = apiAvailablePort !== apiPreferredPort
+  }
+
+  // Check web port
+  if (side.includes('web')) {
+    // Extract any ports the user forwarded to the webpack server and prefer that instead
+    const forwardedPortMatches = [
+      ...forward.matchAll(/\-\-port(\=|\s)(?<port>[^\s]*)/g),
+    ]
+    if (forwardedPortMatches.length) {
+      webPreferredPort = forwardedPortMatches.pop().groups.port
+    }
+
+    webAvailablePort = await getFreePort(webPreferredPort, [
+      apiPreferredPort,
+      apiAvailablePort,
+    ])
+    if (webAvailablePort === -1) {
+      errorTelemetry(
+        process.argv,
+        `Error could not determine a free port for the web server`
+      )
+      console.error(`Error could not determine a free port for the web server`)
+      process.exit(-1)
```

Comment:
(See another similar suggestion)

```suggestion
      process.exit(1)
```